### ISSUE TITLE
[Phase 2] Neural Network Architecture

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/feynman-kac-pinn#3]
## Summary

Implement the neural network architecture that learns the PDE solution u(x) from Monte Carlo estimates. The network takes spatial coordinates as input and outputs the solution value.

**Parent Issue:** #1
**Depends On:** #2 (Phase 1: Random Walk Engine)

## Learning Context

Before coding, understand:
- **MLP Architecture**: Fully connected layers with nonlinear activations
- **Xavier/He Initialization**: Weight init schemes for stable training
- **Residual Connections**: Skip connections for deeper networks (optional)

## Files to Create

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `ml/models/pinn.py` | ~150 | Create | Main PINN architecture |
| `ml/models/activations.py` | ~50 | Create | Custom activation functions |
| `ml/training/trainer.py` | ~200 | Create | Training loop |
| `ml/training/losses.py` | ~50 | Create | Loss functions |
| `ml/training/schedulers.py` | ~80 | Create | LR schedulers |
| `tests/test_pinn.py` | ~100 | Create | Network tests |

### Task 1: Network Architecture
- [ ] Create `FeynmanKacPINN` class with configurable layers
- [ ] Add `tanh`, `swish`, `gelu` activation options
- [ ] Implement Xavier and He initialization
- [ ] Add optional residual connections for deep networks

### Task 2: Training Components
- [ ] Create `FeynmanKacTrainer` class
- [ ] Implement MSE loss between network and MC estimates
- [ ] Add gradient clipping option
- [ ] Implement early stopping based on validation loss

### Task 3: Learning Rate Scheduling
- [ ] Implement cosine annealing scheduler
- [ ] Add warmup period option
- [ ] Create reduce-on-plateau scheduler

### Task 4: Checkpointing
- [ ] Save/load model weights
- [ ] Save optimizer state for resumption
- [ ] Track training history (loss, LR, grad norm)

### Task 5: Device Management
- [ ] Auto-detect MPS/CUDA/CPU
- [ ] Handle device transfers correctly
- [ ] Test on all available backends

## Code Skeleton

```python
# ml/models/pinn.py
import torch
import torch.nn as nn
from typing import List, Optional

class FeynmanKacPINN(nn.Module):
    """
    MLP that approximates PDE solution u(x).

    Architecture:
        Input(dim) -> [Dense(width) -> Activation]×depth -> Output(1)
    """

    def __init__(
        self,
        input_dim: int,
        hidden_dims: List[int] = [64, 64, 64, 64],
        activation: str = "tanh",
        output_activation: Optional[str] = None,
        use_residual: bool = False,
    ):
        super().__init__()
        self.input_dim = input_dim
        self.use_residual = use_residual

        # Build layers
        layers = []
        prev_dim = input_dim
        for hidden_dim in hidden_dims:
            layers.append(nn.Linear(prev_dim, hidden_dim))
            layers.append(self._get_activation(activation))
            prev_dim = hidden_dim

        self.hidden = nn.Sequential(*layers)
        self.output = nn.Linear(prev_dim, 1)

        if output_activation:
            self.output_act = self._get_activation(output_activation)
        else:
            self.output_act = None

        self._init_weights()

    def _get_activation(self, name: str) -> nn.Module:
        activations = {
            "tanh": nn.Tanh(),
            "relu": nn.ReLU(),
            "gelu": nn.GELU(),
            "swish": nn.SiLU(),
        }
        return activations[name]

    def _init_weights(self):
        for m in self.modules():
            if isinstance(m, nn.Linear):
                nn.init.xavier_uniform_(m.weight)
                nn.init.zeros_(m.bias)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        h = self.hidden(x)
        out = self.output(h)
        if self.output_act:
            out = self.output_act(out)
        return out
```

```python
# ml/training/trainer.py
class FeynmanKacTrainer:
    """Training orchestrator for Feynman-Kac PINN."""

    def __init__(
        self,
        model: FeynmanKacPINN,
        problem,
        lr: float = 1e-3,
        device: str = None,
    ):
        self.model = model
        self.problem = problem
        self.device = device or get_device()
        self.model.to(self.device)
        self.optimizer = torch.optim.Adam(model.parameters(), lr=lr)
        self.history = {"loss": [], "lr": []}

    def train_step(self, batch_size: int, n_mc_paths: int) -> dict:
        """Single training step."""
        self.model.train()

        # Sample interior points
        x = self.problem.domain.sample_interior(batch_size, self.device)

        # Compute MC estimates (ground truth)
        mc_estimates, _ = feynman_kac_estimate(
            x, self.problem.boundary_condition,
            self.problem.domain, self.problem.potential,
            n_paths=n_mc_paths
        )

        # Forward pass
        predictions = self.model(x).squeeze(-1)

        # Loss
        loss = F.mse_loss(predictions, mc_estimates)

        # Backward
        self.optimizer.zero_grad()
        loss.backward()
        self.optimizer.step()

        return {"loss": loss.item()}
```

## Verification Criteria

| Test | Expected | Tolerance |
|------|----------|-----------|
| Forward shape | (batch, 1) | Exact |
| Overfit single point | loss < 1e-6 | Exact |
| 2D Laplace on square | <5% error | L2 norm |
| Training converges | monotonic decrease | 100 steps |

## Definition of Done

- [ ] Network trains on 2D and 10D problems
- [ ] Checkpoints save/load correctly
- [ ] Training on MPS achieves >100 steps/second
- [ ] All tests pass

**Comment by Sakeeb91** [comment]
Phase 2 is complete and merged to main in 30 atomic commits (ending at 2ad5bd5). This adds the PINN architecture stack, activation/init modules, trainer/loss/scheduler components, checkpointing, and comprehensive tests.
